### DISCUSSION TITLE
Mouse Button Clicks behave differently in different systems

### DISCUSSION THREAD

**Discussion by pubrrr** [bevyengine/bevy#3712] [Q&A] [answered]
I experience (at least for me) unexpected behavior with clicks: I have a menu, upon clicking a button I start the game by changing  `State<AppState>` from `Menu` to `InGame`. Then I want to watch for clicks via `Res<Input<MouseButton>>` in the system `watch_clicks`.
- If I use `add_system_set(SystemSet::on_update(AppState::InGame).with_system(watch_clicks))`, then the click on the button in the menu is still recognized as `just_pressed`. I would actually expect the contrary.
- If I use `add_system(watch_clicks)`, then it behaves as I would expect: `just_pressed(MouseButton::Left)` is false. However I don't want to unnecessarily run this system in the menu as I don't need it there.

Otherwise the code is the same. Can anyone explain why this happens?

I reproduced the behavior in this example (Bevy 0.6):
```
use bevy::DefaultPlugins;
use bevy::prelude::{
    App, ButtonBundle, Color, Commands, Input, Interaction, MouseButton, Query, Res, ResMut, Size,
    State, Style, SystemSet, UiCameraBundle, UiColor, Val, World
};

#[derive(Clone, Hash, Debug, Eq, PartialEq)]
enum MyState {
    Menu,
    Game,
}

fn main() {
    App::new()
        .add_state(MyState::Menu)
        .add_plugins(DefaultPlugins)
        .add_startup_system(insert_button)
        .add_system_set(
            SystemSet::on_update(MyState::Menu).with_system(watch_menu_clicks)
        )
        .add_system_set( // will panic with "Res(Input { pressed: {Left}, just_pressed: {Left}, just_released: {} })"
            SystemSet::on_update(MyState::Game).with_system(watch_game_clicks)
        )
        // .add_system(watch_game_clicks) // will panic with "Res(Input { pressed: {Left}, just_pressed: {}, just_released: {} })"
        .run();
}

fn insert_button(mut commands: Commands) {
    commands.spawn_bundle(ButtonBundle {
        color: UiColor(Color::BLUE),
        style: Style {
            size: Size::new(Val::Px(50.), Val::Px(50.)),
            ..Default::default()
        },
        ..Default::default()
    });
    commands.spawn_bundle(UiCameraBundle::default());
}

fn watch_menu_clicks(
    mut state: ResMut<State<MyState>>,
    query: Query<&Interaction>,
) {
    match query.single() {
        Interaction::Clicked => state.set(MyState::Game).unwrap(),
        _ => {}
    };
}

fn watch_game_clicks(mouse: Res<Input<MouseButton>>, state: Res<State<MyState>>) {
    if state.current() == &MyState::Game {
        panic!("{:?}", mouse);
    }
}
```

**Comment by DJMcNab** [comment, marked as answer]
If you add the `ReportExecutionOrderAmbiguities` resource to your world, you'll see that `watch_menu_clicks` and `watch_game_clicks` are ambiguous in both cases. By coincidence of how the executor's ordering runs, it happens that when it's not using the state handling, `watch_game_clicks` runs first in a frame; when you are using the state version, `watch_menu_clicks` happens first. 

Hence, when using states, the `watch_game_clicks` runs in the same frame as you switch to being in game, so the button was `just_pressed` in that frame. When not using states, `watch_game_clicks` runs the frame after, whilst the button was still being pressed, but it's 'just_pressed'-ness had been cleared.

**Reply by pubrrr** [reply]
Thank you. Weird coincidence, but makes sense. 

How do I best resolve this? Probably by running `watch_menu_clicks` `.after` the `watch_game_clicks`? Concerning the actual game logic, they are pretty independent, so I feel like I should actually introduce no dependency here...

**Reply by DJMcNab** [reply]
Unfortunately I don't have a good answer for you here. I think you would have to add a running order dependency though, yeah.

**Reply by hymm** [reply]
you might want to manually call clear on the input
https://docs.rs/bevy/latest/bevy/input/struct.Input.html#method.clear

**Reply by pubrrr** [reply]
Clearing the `Input` was my workaround before I knew what the actual issue was. Neither of of them is really satisfying, I'll have to think what I like more.
